### PR TITLE
Document enc_effects.c

### PR DESCRIPTION
**Pull request by lhearachel**
[External research](https://gist.github.com/lhearachel/ade4bad51d4836247b9310f137a56abd)

## Renamed Files

* `include/unk_02051B50.h` -> `include/enc_effects.h`
* `include/overlay006/struct_ov6_02240D5C.h` -> `include/overlay006/battle_params.h`
* `include/struct_defs/struct_0207A9CC.h` -> `include/struct_defs/trainer_data.h`
* `src/unk_02051B50.c` -> `src/enc_effects.c`

## Renamed Structs and Fields

* `UnkStruct_020EC208` -> `EncEffectsDataPair`
  * `UnkStruct_020EC208::unk_00` -> `EncEffectsDataPair::cutInEffect`
  * `UnkStruct_020EC208::unk_02` -> `EncEffectsDataPair::sdatBGMusic`
* `UnkStruct_ov6_02240D5C` -> `BattleParams`
  * `UnkStruct_ov6_02240D5C::unk_00` -> `BattleParams::battleType`
  * `UnkStruct_ov6_02240D5C::unk_02` -> `BattleParams::parties`
  * `UnkStruct_ov6_02240D5C::unk_28` -> `BattleParams::trainerData`
  * `UnkStruct_ov6_02240D5C::unk_134` -> `BattleParams::mapHeaderID`
* `UnkStruct_0207A9CC` -> `TrainerData`
  * `UnkStruct_0207A9CC::unk_01` -> `TrainerData::class`

## Renamed Functions

* `sub_02051B50` -> `EncEffects_GetEffectPair`
* `sub_02051BBC` -> `EncEffects_CutInEffectForPair`
* `sub_02051BE8` -> `EncEffects_BGMForPair`
* `sub_02051C00` -> `EncEffects_CutInEffect`
* `sub_02051C10` -> `EncEffects_BGM`
* `sub_02051C20` -> `EncEffects_TrainerClassEffect`
* `sub_02051CD0` -> `EncEffects_WildPokemonEffect`

## Related Changes

* `ov5_021DEEC8` -> `CutInEffects_ForBattle`
* `sub_02054A40` -> `Party_FirstBattler`

### CODE REVIEW

File: src/enc_effects.c [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+    [ENCEFF_DOUBLE_WILD]        = {ENCEFF_CUTIN_DOUBLE,            SEQ_BATTLE_WILD_POKEMON},
+    [ENCEFF_FRONTIER_BRAIN]     = {ENCEFF_CUTIN_FRONTIER,          SEQ_BATTLE_FRONTIER_BRAIN},
+    [ENCEFF_DOUBLE_LEADER]      = {ENCEFF_CUTIN_DOUBLE,            SEQ_BATTLE_GYM_LEADER},
+    [ENCEFF_NORMAL_TRAINER]     = {ENCEFF_CUTIN_USE_LOCAL,         SEQ_BATTLE_TRAINER},
+    [ENCEFF_NORMAL_WILD]        = {ENCEFF_CUTIN_USE_LOCAL,         SEQ_BATTLE_WILD_POKEMON}
+};
+
+static u32 EncEffects_GetEffectPair(const UnkStruct_ov6_02240D5C *battleParams);
+static u32 EncEffects_CutInEffectForPair(u32 param0, const UnkStruct_ov6_02240D5C * param1);
+static u32 EncEffects_BGMForPair(u32 param0, const UnkStruct_ov6_02240D5C * param1);
```
Comment (lhearachel) [inline src/enc_effects.c:114]: I'm not super happy with these names, and I'm open to suggestions for them.

My one caveat is that I wanted `EncEffects_CutInEffect` and `EncEffects_BGM` to be the two external functions.

File: src/enc_effects.c [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+{
+    return EncEffects_CutInEffectForPair(EncEffects_GetEffectPair(battleParams), battleParams);
+}
+
+u32 EncEffects_BGM (const UnkStruct_ov6_02240D5C *battleParams)
+{
+    return EncEffects_BGMForPair(EncEffects_GetEffectPair(battleParams), battleParams);
+}
+
+static u32 EncEffects_TrainerClassEffect (u32 trainerClass)
```
Comment (lhearachel) [inline src/enc_effects.c:210]: I tried to simplify both this function and the one below for wild Pokemon in order to get rid of the `result` declaration at the top and instead issue a `return` for each `case`. That, unfortunately, didn't match, so here we are.

File: include/constants/battle.h [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+#define BATTLE_TYPE_DOUBLES         (0x00000002)
+#define BATTLE_TYPE_LINK            (0x00000004)
+#define BATTLE_TYPE_2vs2            (0x00000008)    // This is, explicitly, two distinct battlers on each side
+#define BATTLE_TYPE_TAG             (0x00000010)    // This is, explicitly, a battle with a partner
+#define BATTLE_TYPE_SAFARI          (0x00000020)
+#define BATTLE_TYPE_AI              (0x00000040)
+#define BATTLE_TYPE_FRONTIER        (0x00000080)
+#define BATTLE_TYPE_ROAMER          (0x00000100)
+#define BATTLE_TYPE_PAL_PARK        (0x00000200)
+#define BATTLE_TYPE_CATCH_TUTORIAL  (0x00000400)
```
Comment (mid-kid) [inline include/constants/battle.h:16]: I can't find a precedent for this in the rest of the codebase, so I'd suggest using the `(1 << N)` notation for bitflags

File: include/constants/sdat.h [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+#define SEQ_BATTLE_CHAMPION         (1122)
+#define SEQ_BATTLE_GALACTIC_GRUNT   (1123)
+#define SEQ_BATTLE_RIVAL            (1124)
+#define SEQ_BATTLE_ARCEUS           (1125)
+#define SEQ_BATTLE_LEGENDARY        (1126)
+#define SEQ_BATTLE_GALACTIC_CMDR    (1134)
+#define SEQ_BATTLE_ELITE_FOUR       (1136)
+#define SEQ_BATTLE_GIRATINA         (1201)
+#define SEQ_BATTLE_FRONTIER_BRAIN   (1202)
+#define SEQ_BATTLE_REGI_TRIO        (1204)
```
[the remark this answers is not in the document]
Comment (mid-kid) [reply include/constants/sdat.h:19]: Raw numbers don't need to be surrounded by braces. Why isn't this an enum?

File: src/enc_effects.c [hunk 5]
```diff
@@ -0,0 +1,339 @@
...
+    ENCEFF_CHAMPION_CYNTHIA,
+
+    ENCEFF_RIVAL,
+
+    ENCEFF_SHAYMIN,
+    ENCEFF_DIALGA_PALKIA,
+    ENCEFF_UXIE_AZELF,
+    ENCEFF_MESPRIT,
+    ENCEFF_ARCEUS,
+    ENCEFF_MINOR_LEGENDARIES,   // Regigigas, Heatran, Darkrai
```
Comment (mid-kid) [inline src/enc_effects.c:51]: Don't align comments

**Review by mid-kid — CHANGES_REQUESTED** [review]

File: include/constants/sdat.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (lhearachel) [reply include/constants/sdat.h:19]: Idea was that it should be an enum eventually, I just did defines out of laziness. I'll make this an enum.

File: include/constants/battle.h (hunk 3 above)
Comment (lhearachel) [reply include/constants/battle.h:16]: There isn't really one right now. Time to set one.

File: src/enc_effects.c [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+static u32 EncEffects_CutInEffectForPair(u32 effectPairID, const UnkStruct_ov6_02240D5C *battleParams);
+static u32 EncEffects_BGMForPair(u32 effectPairID, const UnkStruct_ov6_02240D5C *battleParams);
+static u32 EncEffects_TrainerClassEffect(u32 trainerClass);
+static u32 EncEffects_WildPokemonEffect(Party *wildParty, int zoneHeaderID);
+
+inline BOOL EncEffects_Galactic(u32 effect)
+{
+    return (effect == ENCEFF_GALACTIC_GRUNT)
+            || (effect == ENCEFF_GALACTIC_CMDR)
+            || (effect == ENCEFF_GALACTIC_CYRUS);
```
Comment (mid-kid) [inline src/enc_effects.c:122]: Braces not necessary.
Not sure about the alignment to the opening bracket here, do we have precedent? I'd just indent the continuation lines by one level...

Comment (lhearachel) [reply src/enc_effects.c:122]: I don't think we have a clear precedent on this one, no. It's not covered in the style guide, so I defaulted to double-indent.

Comment (mid-kid) [reply src/enc_effects.c:122]: Ah, double indent is fine, will work for if() blocks as well.

File: src/enc_effects.c [hunk 7]
```diff
@@ -0,0 +1,339 @@
...
+}
+
+static u32 EncEffects_GetEffectPair (const UnkStruct_ov6_02240D5C *battleParams)
+{
+    u32 battleType = battleParams->unk_00;
+
+    if (battleType & BATTLE_TYPE_TRAINER) {
+        u32 trainerEffect = EncEffects_TrainerClassEffect(battleParams->unk_28[1].unk_01);
```
Comment (mid-kid) [inline src/enc_effects.c:130]: Might be a good idea to name these struct fields in a followup commit.

**Review by mid-kid — APPROVED** [review]

**Pull request merged**